dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,713 @@
+           of 3 eigenvalues, the volume fraction of the free water
+           compartment, and the estimate of the non diffusion-weighted signal
+           S0). If the initial fw_paramters are not given, function will use
+           the WLS free water elimination algorithm to estimate the parameters
+           first guess.
+           Default: None
+
+    Diso : float, optional
+        Value of the free water isotropic diffusion. Default is set to 3e-3
+        $mm^{2}.s^{-1}$. Please ajust this value if you are assuming different
+        units of diffusion.
+
+    weighting: str, optional
+           the weighting scheme to use in considering the
+           squared-error. Default behavior is to use uniform weighting. Other
+           options: 'sigma' 'gmm'
+
+    sigma: float, optional
+        If the 'sigma' weighting scheme is used, a value of sigma needs to be
+        provided here. According to [Chang2005]_, a good value to use is
+        1.5267 * std(background_noise), where background_noise is estimated
+        from some part of the image known to contain no signal (only noise).
+
+    cholesky : bool, optional 
+        If true it uses cholesky decomposition to insure that diffusion tensor
+        is positive define.
+        Default: False
+
+    f_transform : bool, optional
+        If true, the water volume fractions is converted during the convergence 
+        procedure to ft = arcsin(2*f - 1) + pi/2, insuring f estimates between
+        0 and 1.
+        Default: True
+
+    Returns
+    -------
+    nlls_params: for each voxel the eigen-values and eigen-vectors of the
+        tissue tensor and the volume fraction of the free water compartment.
+    """
+    # Flatten for the iteration over voxels:
+    flat_data = data.reshape((-1, data.shape[-1]))
+
+    # Use the WLS method parameters as the starting point if fw_params is None:
+    if fw_params==None:
+        fw_params = wls_fit_tensor(design_matrix, flat_data,  Diso=Diso)
+
+    fw_params = fw_params.reshape((-1, fw_params.shape[-1]))
+    
+    for vox in range(flat_data.shape[0]):
+        if np.all(flat_data[vox] == 0):
+            raise ValueError("The data in this voxel contains only zeros")
+
+        params = fw_params[vox]
+
+        # converting evals and evecs to diffusion tensor elements
+        evals = params[:3]
+        evecs = params[3:12].reshape((3, 3))
+        dt = lower_triangular(np.dot(np.dot(evecs, np.diag(evals)), evecs.T))
+        s0 = params[13]
+
+        # Cholesky decomposition if requested
+        if cholesky:
+            dt = lower_triangular_to_cholesky(dt)
+
+        # f transformation if requested
+        if f_transform:
+            f = np.arcsin(2*params[12] - 1) + np.pi/2
+        else:
+            f = params[12]
+
+        # Use the Levenberg-Marquardt algorithm wrapped in opt.leastsq
+        start_params = np.concatenate((dt, [-np.log(s0), f]), axis=0)
```

Comment:
Copy the parameters so we have them later without needing to re-create the vector.
```
input_params = params  # ETP copy it in case we need it later
```